esphome/components/esp_adf/microphone/esp_adf_microphone.cpp

```diff
@@ -0,0 +1,336 @@
+#include "esp_adf_microphone.h"
+
+#ifdef USE_ESP_IDF
+
+#include <driver/i2s.h>
+
+#include "esphome/core/hal.h"
+#include "esphome/core/log.h"
+
+#include <algorithm_stream.h>
+#include <audio_element.h>
+#include <audio_hal.h>
+#include <audio_pipeline.h>
+#include <filter_resample.h>
+#include <i2s_stream.h>
+#include <raw_stream.h>
+#include <recorder_sr.h>
+
+#include <board.h>
+
+namespace esphome {
+namespace esp_adf {
+
+static const char *const TAG = "esp_adf.microphone";
+
+void ESPADFMicrophone::setup() {
+  this->ring_buffer_ = rb_create(8000, sizeof(int16_t));
+  if (this->ring_buffer_ == nullptr) {
+    ESP_LOGW(TAG, "Could not allocate ring buffer.");
+    this->mark_failed();
+    return;
+  }
+
+  this->read_event_queue_ = xQueueCreate(20, sizeof(TaskEvent));
+  this->read_command_queue_ = xQueueCreate(20, sizeof(CommandEvent));
+}
+
+void ESPADFMicrophone::start() {
+  if (this->is_failed())
+    return;
+  if (this->state_ == microphone::STATE_STOPPING) {
+    ESP_LOGW(TAG, "Microphone is stopping, cannot start.");
+    return;
+  }
+  this->state_ = microphone::STATE_STARTING;
+}
+void ESPADFMicrophone::start_() {
+  if (!this->parent_->try_lock()) {
+    return;
+  }
+
+  xTaskCreate(ESPADFMicrophone::read_task, "read_task", 8192, (void *) this, 0, &this->read_task_handle_);
+}
+
+void ESPADFMicrophone::read_task(void *params) {
+  ESPADFMicrophone *this_mic = (ESPADFMicrophone *) params;
+  TaskEvent event;
+
+  ExternalRAMAllocator<int16_t> allocator(ExternalRAMAllocator<int16_t>::ALLOW_FAILURE);
+  int16_t *buffer = allocator.allocate(BUFFER_SIZE / sizeof(int16_t));
+  if (buffer == nullptr) {
+    event.type = TaskEventType::WARNING;
+    event.err = ESP_ERR_NO_MEM;
+    xQueueSend(this_mic->read_event_queue_, &event, portMAX_DELAY);
+
+    event.type = TaskEventType::STOPPED;
+    event.err = ESP_OK;
+    xQueueSend(this_mic->read_event_queue_, &event, portMAX_DELAY);
+
+    while (true) {
+      delay(10);
+    }
+    return;
+  }
+
+  event.type = TaskEventType::STARTING;
+  xQueueSend(this_mic->read_event_queue_, &event, portMAX_DELAY);
+
+  audio_pipeline_cfg_t pipeline_cfg = {
+      .rb_size = 8 * 1024,
+  };
+  audio_pipeline_handle_t pipeline = audio_pipeline_init(&pipeline_cfg);
+
+  i2s_driver_config_t i2s_config = {
+      .mode = (i2s_mode_t) (I2S_MODE_MASTER | I2S_MODE_RX),
+      .sample_rate = 16000,
+      .bits_per_sample = I2S_BITS_PER_SAMPLE_16BIT,
+      .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,
+      .communication_format = I2S_COMM_FORMAT_STAND_I2S,
+      .intr_alloc_flags = ESP_INTR_FLAG_LEVEL2 | ESP_INTR_FLAG_IRAM,
+      .dma_buf_count = 8,
+      .dma_buf_len = 128,
+      .use_apll = false,
+      .tx_desc_auto_clear = true,
+      .fixed_mclk = 0,
+      .mclk_multiple = I2S_MCLK_MULTIPLE_256,
+      .bits_per_chan = I2S_BITS_PER_CHAN_DEFAULT,
+  };
+
+  i2s_stream_cfg_t i2s_cfg = {
+      .type = AUDIO_STREAM_READER,
+      .i2s_config = i2s_config,
+      .i2s_port = I2S_NUM_0,
```

Comment:
```suggestion
      .i2s_port = static_cast<i2s_port_t>(CODEC_ADC_I2S_PORT),
```
This is required for boards that use a mic that is not on I2S 0, like Korvo1